tests/utils/managed_process.py

```diff
@@ -0,0 +1,239 @@
+# You may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import logging
+import os
+import shutil
+import socket
+import subprocess
+import time
+from dataclasses import dataclass, field
+from typing import Any, List, Optional
+
+import psutil
+import requests
+
+
+@dataclass
+class ManagedProcess:
+    command: List[str]
+    env: Optional[dict] = None
+    health_check_ports: List[int] = field(default_factory=list)
+    health_check_urls: List[Any] = field(default_factory=list)
+    timeout: int = 300
+    working_dir: Optional[str] = None
+    display_output: bool = False
+    data_dir: Optional[str] = None
+    terminate_existing: bool = True
+    stragglers: List[str] = field(default_factory=list)
+    log_dir: str = os.getcwd()
+
+    _logger = logging.getLogger()
+    _command_name = None
+    _log_path = None
+    _tee_proc = None
+    _sed_proc = None
+
+    def __enter__(self):
+        try:
+            self._logger = logging.getLogger(self.__class__.__name__)
+            self._command_name = self.command[0]
+            os.makedirs(self.log_dir, exist_ok=True)
+            log_name = f"{self._command_name}.log.txt"
+            self._log_path = os.path.join(self.log_dir, log_name)
+
+            if self.data_dir:
+                self._remove_directory(self.data_dir)
+
+            self._terminate_existing()
+            self._start_process()
+            self.timeout -= self._check_ports(self.timeout)
+            self.timeout -= self._check_urls(self.timeout)
+
+            return self.proc
+
+        except Exception as e:
+            self.__exit__(None, None, None)
+            raise e
+
+    def __exit__(self, exc_type, exc_val, exc_tb):
+        process_list = [self.proc, self._tee_proc, self._sed_proc]
+        for process in process_list:
+            if process:
+                if process.stdout:
+                    process.stdout.close()
+                if process.stdin:
+                    process.stdin.close()
+                self._terminate_process_tree(process.pid)
+                process.wait()
+        if self.data_dir:
+            self._remove_directory(self.data_dir)
+
+        for ps_process in psutil.process_iter(["name", "cmdline"]):
+            if ps_process.name() in self.stragglers:
+                self._terminate_process_tree(ps_process.pid)
+
+    def _start_process(self):
+        assert self._command_name
+        assert self._log_path
+
+        self._logger.info(
+            f"Running command: {' '.join(self.command)} in {self.working_dir or os.getcwd()}"
+        )
+
+        stdin = subprocess.DEVNULL
```

Comment:
wait wut? this is confusing

why is standard in /dev/null, standard output a random pipe, and standard error mapped to standard output?

😕

Reply:
but you do understand that stderr and stdout have different behaviors and if the running process is depending on stderr behavior (i.e. expecting the pipe to flush itself), it could lead to unexpected results?

if so, and that's fine - feel free to resolve this comment.

Reply:
I think its fine for now .... it should behave similarly to redirect in bash - if we need to separate the two we can - but today mostly expectation is for both to be in a combined log - IIUC